File: js/android/src/main/java/com/reactnativesecurekeystore/RNModule/SecureKeystoreModule.java
package com.reactnativesecurekeystore;

import com.reactnativesecurekeystore.SecureKeystoreImpl;
import com.reactnativesecurekeystore.KeyGeneratorImpl;
import com.reactnativesecurekeystore.CipherBoxImpl;
import android.app.Activity;
import com.reactnativesecurekeystore.DeviceCapability;
import android.content.Intent;
import androidx.annotation.NonNull;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.reactnativesecurekeystore.biometrics.Biometrics;
import com.reactnativesecurekeystore.common.Util;
import kotlin.Unit;
import android.content.Context;
import org.jetbrains.annotations.Nullable;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;
import androidx.fragment.app.FragmentActivity;
import com.facebook.react.bridge.ActivityEventListener;

public class SecureKeystoreModule extends ReactContextBaseJavaModule implements ActivityEventListener {
    private final KeyGeneratorImpl keyGenerator = new KeyGeneratorImpl();
    private final CipherBoxImpl cipherBox = new CipherBoxImpl();
    private final Biometrics biometrics;
    private final SecureKeystoreImpl keystore;
    private final DeviceCapability deviceCapability;
    private final String logTag;

    private ReactApplicationContext context;

    public SecureKeystoreModule(ReactApplicationContext reactContext) {
        super(reactContext);
        context = reactContext;
        Context context = getReactApplicationContext();
        this.biometrics = new Biometrics();
        this.keystore = new SecureKeystoreImpl(keyGenerator, cipherBox, biometrics);
        this.deviceCapability = new DeviceCapability(keystore, keyGenerator, biometrics);
        this.logTag = Util.Companion.getLogTag(getClass().getSimpleName());
    }

    @Override
    public String getName() {
        return "SecureKeystoreModule";
    }

    @Override
    public void onNewIntent(Intent intent) {
        // Handle new intent here if needed
    }

    @Override
    public void onActivityResult(Activity activity, int requestCode, int resultCode, @Nullable Intent data) {
        // Handle activity result here if needed
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public boolean deviceSupportsHardware() {
        return deviceCapability.supportsHardwareKeyStore();
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public boolean hasAlias(String alias) {
        return keystore.hasAlias(alias);
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public void updatePopup(String title, String description) {
        Biometrics.Companion.updatePopupDetails(title, description);
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public void generateKey(String alias, boolean isAuthRequired, Integer authTimeout) {
        keystore.generateKey(alias, isAuthRequired, authTimeout);
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String generateKeyPair(String alias, boolean isAuthRequired, Integer authTimeout) {
        return keystore.generateKeyPair(alias, isAuthRequired, authTimeout);
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public String generateKeyPairEC(String alias, boolean isAuthRequired, Integer authTimeout) {
        return keystore.generateKeyPairEC(alias, isAuthRequired, authTimeout);
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public void generateHmacshaKey(String alias) {
        keystore.generateHmacSha256Key(alias);
    }

    @ReactMethod
    public void encryptData(String alias, String data, Promise promise) {
        Function1<String, Unit> successLambda = new Function1<String, Unit>() {
            @Override
            public Unit invoke(String encryptedText) {
                promise.resolve(encryptedText);
                return Unit.INSTANCE;
            }
        };

        Function2<Integer, String, Unit> failureLambda = new Function2<Integer, String, Unit>() {
            @Override
            public Unit invoke(Integer code, String message) {
                promise.reject(code.toString(), message);
                return Unit.INSTANCE;
            }
        };
        keystore.encryptData(
                alias,
                data,
                successLambda,
                failureLambda,
                context.getCurrentActivity()
        );
    }

    @ReactMethod
    public void decryptData(String alias, String encryptedText, Promise promise) {
        Function1<String, Unit> successLambda = new Function1<String, Unit>() {
            @Override
            public Unit invoke(String data) {
                promise.resolve(data);
                return Unit.INSTANCE;
            }
        };

        Function2<Integer, String, Unit> failureLambda = new Function2<Integer, String, Unit>() {
            @Override
            public Unit invoke(Integer code, String message) {
                promise.reject(code.toString(), message);
                return Unit.INSTANCE;
            }
        };

        keystore.decryptData(alias, encryptedText, successLambda, failureLambda, context.getCurrentActivity());
    }

    @ReactMethod
    public void generateHmacSha(String alias, String data, Promise promise) {
        Function1<String, Unit> successLambda = new Function1<String, Unit>() {
            @Override
            public Unit invoke(String sha) {
                promise.resolve(sha);
                return Unit.INSTANCE;
            }
        };

        Function2<Integer, String, Unit> failureLambda = new Function2<Integer, String, Unit>() {
            @Override
            public Unit invoke(Integer code, String message) {
                promise.reject(code.toString(), message);
                return Unit.INSTANCE;
            }
        };

        keystore.generateHmacSha(
                alias,
                data,
                successLambda,
                failureLambda
        );
    }

    @ReactMethod
    public void sign(String signAlgorithm, String alias, String data, Promise promise) {
        Function1<String, Unit> successLambda = new Function1<String, Unit>() {
            @Override
            public Unit invoke(String signature) {
                promise.resolve(signature);
                return Unit.INSTANCE;
            }
        };

        Function2<Integer, String, Unit> failureLambda = new Function2<Integer, String, Unit>() {
            @Override
            public Unit invoke(Integer code, String message) {
                promise.reject(code.toString(), message);
                return Unit.INSTANCE;
            }
        };

        keystore.sign(
                signAlgorithm,
                alias,
                data,
                successLambda,
                failureLambda,
                context.getCurrentActivity()
        );
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public void clearKeys() {
        keystore.removeAllKeys();
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public boolean hasBiometricsEnabled() {
        return deviceCapability.hasBiometricsEnabled(context.getCurrentActivity());
    }
}

File: android/src/main/java/com/reactnativesecurekeystore/RNModule/SecureKeystoreModule.kt
package com.reactnativesecurekeystore

import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.reactnativesecurekeystore.biometrics.Biometrics
import com.reactnativesecurekeystore.common.Util
import kotlin.math.log

class SecureKeystoreModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {
  private val keyGenerator = KeyGeneratorImpl()
  private val cipherBox = CipherBoxImpl()
  private val biometrics = Biometrics(reactContext)
  private val keystore = SecureKeystoreImpl(keyGenerator, cipherBox, biometrics)
  private val deviceCapability = DeviceCapability(keystore, keyGenerator, biometrics)
  private val logTag = Util.getLogTag(javaClass.simpleName)

  override fun getName(): String {
    return "SecureKeystore"
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun deviceSupportsHardware(): Boolean {
    val supportsHardware = deviceCapability.supportsHardwareKeyStore()
    Log.d(logTag, "Device supports Hardware $supportsHardware")
    return supportsHardware
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun hasAlias(alias: String): Boolean {
    val exists = keystore.hasAlias(alias)
    Log.d(logTag, "Alias Exist $exists")
    return exists
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun updatePopup(title: String, description: String) {
    Biometrics.updatePopupDetails(title, description)
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun generateKey(alias: String, isAuthRequired: Boolean, authTimeout: Int?) {
    Log.d(logTag, "Generating a key for $alias")

    keystore.generateKey(alias, isAuthRequired, authTimeout)
  }

  // Generates KeyPair and returns Public key
  @ReactMethod(isBlockingSynchronousMethod = true)
  fun generateKeyPair(alias: String, isAuthRequired: Boolean, authTimeout: Int?): String {
    Log.d(logTag, "Generating a keyPair")
    return keystore.generateKeyPair(alias, isAuthRequired, authTimeout)
  }

  // Generates hmacsha256 key
  @ReactMethod(isBlockingSynchronousMethod = true)
  fun generateHmacshaKey(alias: String) {
    Log.d(logTag, "Generating a generateHmacsha256 Key")
    keystore.generateHmacSha256Key(alias)
  }

  @ReactMethod
  fun encryptData(alias: String, data: String, promise: Promise) {
    Log.d(logTag, "Encrypting data")

    keystore.encryptData(
      alias,
      data,
      onSuccess = { encryptedText -> run { promise.resolve(encryptedText) } },
      onFailure = { code, message -> run { promise.reject(code.toString(), message) } })
  }

  @ReactMethod
  fun decryptData(alias: String, encryptedText: String, promise: Promise) {
    Log.d(logTag, "decrypting data with $alias")

    keystore.decryptData(
      alias,
      encryptedText,
      onSuccess = { data -> run { promise.resolve(data) } },
      onFailure = { code, message -> run { promise.reject(code.toString(), message) } })
  }

  @ReactMethod
  fun generateHmacSha(alias: String, data: String, promise: Promise) {
    Log.d(logTag, "generating HMAC Sha for data")

    keystore.generateHmacSha(alias, data,
      onSuccess = { sha -> run { promise.resolve(sha) } },
      onFailure = { code, message -> run { promise.reject(code.toString(), message) } })
  }

  @ReactMethod
  fun sign(signAlgoritm: String, alias: String, data: String, promise: Promise) {
    Log.d(logTag, "signing data")

    keystore.sign(
      signAlgoritm,
      alias,
      data,
      onSuccess = { signature -> run { promise.resolve(signature) } },
      onFailure = { code, message -> run { promise.reject(code.toString(), message) } })
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun clearKeys() {
    keystore.removeAllKeys()
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun hasBiometricsEnabled(): Boolean {
    val isEnabled = deviceCapability.hasBiometricsEnabled()
    Log.d(logTag, "Device biometrics enabled -> $isEnabled")
    return isEnabled
  }
}
